core.max_cache_size is ignored on linux

You can set core.max_cache_size, but Vapoursynth seemingly ignores it, so that the memory usage rises above the threshold without max_cache_size being enforced.

Under Windows, it also seems to be ignored, but only until about "max + 500MB", after which vapoursynth is cleaning up itself.

Arch Linux: 
vapoursynth-git
descale-git

test script:
https://pastebin.com/h0DkxxU4
You can use the test script directly in the Python shell.
Use the process command 2-3 times and your RAM usage should triple the max_cache_size. With each following execution, RAM usage continues to increase.

```
/usr/bin/ld: src/vspipe/vspipe.o: in function `outputFrame(VSFrameRef const*, VSPipeOutputData*)':
vspipe.cpp:(.text+0x15d8): undefined reference to `MD5_Update'
/usr/bin/ld: vspipe.cpp:(.text+0x177e): undefined reference to `MD5_Update'
/usr/bin/ld: src/vspipe/vspipe.o: in function `main':
vspipe.cpp:(.text.startup+0xb36): undefined reference to `MD5_Init'
/usr/bin/ld: vspipe.cpp:(.text.startup+0xba0): undefined reference to `MD5_Final'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `forkpty'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `openpty'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `pthread_create'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `pthread_key_create'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `sem_wait'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `sem_init'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `pthread_getspecific'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `sem_destroy'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `dlopen'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `pthread_key_delete'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `pthread_attr_setstacksize'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `pthread_kill'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `sem_post'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `sem_trywait'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `dlerror'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `dlsym'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `sem_timedwait'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `pthread_sigmask'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `pthread_getcpuclockid'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `pthread_setspecific'
/usr/bin/ld: ./.libs/libvapoursynth-script.so: undefined reference to `pthread_detach'
collect2: error: ld returned 1 exit status
make: *** [Makefile:1342: vspipe] Error 1
```

--- a/src/cython/vapoursynth.pyx
+++ b/src/cython/vapoursynth.pyx
@@ -43,6 +43,7 @@ import weakref
 import atexit
 import contextlib
 import logging
+import functools
 from threading import local as ThreadLocal, Lock, RLock
 from types import MappingProxyType
 from collections import namedtuple
@@ -643,18 +644,20 @@ cdef class CallbackData(object):
     def for_future(self, object future, object wrap_call=None):
         if wrap_call is None:
             wrap_call = lambda func, *args, **kwargs: func(*args, **kwargs)
-        self.callback = self.handle_future
+        self.callback = functools.partial(CallbackData.handle_future,
+                                          self.env, future, wrap_call)
         self.future = future
         self.wrap_cb = wrap_call
 
-    def handle_future(self, node, n, result):
+    @staticmethod
+    def handle_future(env, future, wrapper, node, n, result):
         if isinstance(result, Error):
-            func = self.future.set_exception
+            func = future.set_exception
         else:
-            func = self.future.set_result
+            func = future.set_result
 
-        with use_environment(self.env).use():
-            self.wrap_cb(func, result)
+        with use_environment(env).use():
+            wrapper(func, result)
 
     def receive(self, n, result):
         self.callback(self.node, n, result)

